File: Coding/python/rate_limiter.py
"""Port of java/RateLimiter.java -> a rate gate (sliding-window semaphore).

Original Java package: code.ds

Java concurrency mapping:
  Semaphore                 -> threading.Semaphore
  ConcurrentLinkedQueue     -> collections.deque guarded by a Lock
  ScheduledExecutorService  -> threading.Timer (self-rescheduling)

The original Java had a bug: a *local* `scheduledPool` shadowed the field, so
the field stayed null and the rescheduling callback threw NullPointerException.
This port fixes that so the gate actually paces requests, and adds shutdown()
so the demo is deterministic & terminating (no lingering timers).

Idea: each WaitToProceed acquires a permit and records an "exit time"
(now + one time unit). A timer releases permits as their exit times come due,
re-arming itself for the next one.
"""

import threading
import time
from collections import deque
from typing import Deque, Optional


def _now_ms() -> int:
    return int(time.monotonic() * 1000)


class RateLimiter:
    def __init__(self, occurrences: int, time_seconds: int):
        if occurrences <= 0:
            raise ValueError("Number of occurrences must be a positive integer")
        self.occurrences = occurrences
        # TimeUnit.MILLISECONDS.convert(time, SECONDS)
        self.time_unit_ms = time_seconds * 1000

        self._semaphore = threading.Semaphore(occurrences)
        self._queue: Deque[int] = deque()
        self._lock = threading.Lock()
        self._shutdown = False
        self._future_task: Optional[threading.Timer] = None

        # Schedule the first exit check one time unit out (like scheduleWithFixedDelay).
        self._arm(self.time_unit_ms / 1000.0)

    def _arm(self, delay_seconds: float) -> None:
        if self._shutdown:
            return
        t = threading.Timer(delay_seconds, self._delayed_task)
        t.daemon = True
        self._future_task = t
        t.start()

    def _delayed_task(self) -> None:
        now = _now_ms()
        with self._lock:
            # Release the semaphore for every exit time that is now due.
            while self._queue and self._queue[0] - now <= 0:
                self._semaphore.release()
                self._queue.popleft()
            if self._queue:
                time_until_next_check = self._queue[0] - now
            else:
                time_until_next_check = self.time_unit_ms
        self.change_scheduler(time_until_next_check)

    def change_scheduler(self, time_ms: float) -> None:
        if self._shutdown:
            return
        if time_ms > 0:
            if self._future_task is not None:
                self._future_task.cancel()
            self._arm(time_ms / 1000.0)

    def wait_to_proceed(self) -> None:
        # Block until we can enter the semaphore, then record our exit time.
        self._semaphore.acquire()
        time_to_exit = _now_ms() + self.time_unit_ms
        with self._lock:
            self._queue.append(time_to_exit)

    def shutdown(self) -> None:
        self._shutdown = True
        with self._lock:
            if self._future_task is not None:
                self._future_task.cancel()


if __name__ == "__main__":
    # Allow 2 occurrences per 1 second; fire 5 requests and watch them get paced.
    gate = RateLimiter(2, 1)
    start = _now_ms()
    for i in range(5):
        gate.wait_to_proceed()
        print("request %d proceeded at +%d ms (approx)" % (i, _now_ms() - start))
    gate.shutdown()
    print("done")

File: Coding/golang/ratelimiter/ratelimiter.go
package main

import (
	"errors"
	"fmt"
	"sync"
	"time"
)

// RateLimiter allows a fixed number of occurrences per unit of time. This is an
// idiomatic Go port of a semaphore-based rate gate: a buffered channel acts as
// the semaphore and a timer returns each permit one time unit after it is taken.
type RateLimiter struct {
	occurrences int
	timeUnit    time.Duration
	permits     chan struct{}
	wg          sync.WaitGroup
}

// NewRateLimiter builds a limiter of `occurrences` per `timeSeconds` seconds.
func NewRateLimiter(occurrences, timeSeconds int) (*RateLimiter, error) {
	if occurrences <= 0 {
		return nil, errors.New("number of occurrences must be a positive integer")
	}
	rl := &RateLimiter{
		occurrences: occurrences,
		timeUnit:    time.Duration(timeSeconds) * time.Second,
		permits:     make(chan struct{}, occurrences),
	}
	for i := 0; i < occurrences; i++ {
		rl.permits <- struct{}{}
	}
	return rl, nil
}

// WaitToProceed blocks until a permit is available. The permit is automatically
// returned after one time unit, which enforces the configured rate.
func (rl *RateLimiter) WaitToProceed() {
	<-rl.permits
	rl.wg.Add(1)
	time.AfterFunc(rl.timeUnit, func() {
		rl.permits <- struct{}{}
		rl.wg.Done()
	})
}

func main() {
	rl, err := NewRateLimiter(3, 1) // 3 occurrences per second
	if err != nil {
		fmt.Println(err)
		return
	}
	start := time.Now()
	for i := 1; i <= 6; i++ {
		rl.WaitToProceed()
		fmt.Printf("request %d allowed at ~%v\n", i, time.Since(start).Round(100*time.Millisecond))
	}
	// Let the outstanding permits be returned so the program exits cleanly.
	rl.wg.Wait()
	fmt.Println("done")
}
